### PR TITLE
Add Packets-per-second benchmark

### PR DESCRIPTION
**Pull request by dPhanekham**
Add a packets-per-second benchmark called netperf_aggregate (this can be changed if needed) which uses netperf UDP_RR between a client VM and two server VMs. Done with a slightly modified version of a script that can be found in https://github.com/HewlettPackard/netperf/blob/master/doc/examples/runemomniaggdemo.sh.

This also required me to change the netperf installation to a slightly newer version so that it would compile with the --enable-demo flag with no errors. I have tested to original netperf benchmark also to assure it still works with no issues.

### CODE REVIEW

File: perfkitbenchmarker/linux_packages/netperf.py [hunk 1]
```diff
@@ -31,46 +30,45 @@
...
-                   './configure --enable-histogram=yes '
-                   '&& make' % (NETPERF_DIR, FLAGS.netperf_histogram_buckets))
 
+  _LoadNetperf(vm)
 
-def _CopyTar(vm):
-  """Copy the tar file for installation.
+  # vm.PushDataFile('netperf.patch', NETLIB_PATCH)
+  # vm.RemoteCommand('cd %s && patch -p2 < netperf.patch' %
+  #                   NETPERF_SRC_DIR)
```
Comment (s-deitz) [inline perfkitbenchmarker/linux_packages/netperf.py]: Remove commented code?

File: perfkitbenchmarker/linux_packages/texinfo.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Module containing build tools installation and cleanup functions."""
+
+
+def _Install(vm):
```
Comment (s-deitz) [inline perfkitbenchmarker/linux_packages/texinfo.py]: I don't believe this function is used or necessary.

File: perfkitbenchmarker/linux_packages/python_rrdtool.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Module containing build tools installation and cleanup functions."""
+
+
+def _Install(vm):
```
Comment (s-deitz) [inline perfkitbenchmarker/linux_packages/python_rrdtool.py]: I don't believe this function is necessary.

File: perfkitbenchmarker/linux_benchmarks/netperf_aggregate_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+
+# INSTALL
+# Ensure gcc, make, automake, texinfo and python-rrdtool are installed
+# Bring a netperf source tree to the systems
+# cd to the top of the tree
+# run ./autogen.sh
+# ./configure --enable-burst --enable-demo --enable-histogram
+# make  # sudo make install if you prefer
+  vm.Install('texinfo')
+  vm.Install('python_rrdtool')
```
Comment (s-deitz) [inline perfkitbenchmarker/linux_benchmarks/netperf_aggregate_benchmark.py]: Why are texinfo and python_rrdtool necessary?

File: perfkitbenchmarker/linux_benchmarks/netperf_aggregate_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+
+def Cleanup(benchmark_spec):
+  """Cleanup netperf on the target vm (by uninstalling).
+
+  Args:
+    benchmark_spec: The benchmark specification. Contains all data that is
+        required to run the benchmark.
+  """
+  vms = benchmark_spec.vms
+  vms[1].RemoteCommand('sudo killall netserver')
```
Comment (s-deitz) [inline perfkitbenchmarker/linux_benchmarks/netperf_aggregate_benchmark.py]: Kill on the second server VM too?

File: perfkitbenchmarker/linux_packages/netperf.py [hunk 6]
```diff
@@ -31,46 +30,45 @@
...
-  try:
-    vm.PushDataFile(NETPERF_TAR, remote_path=(INSTALL_DIR + '/'))
-  except ResourceNotFound:
-    vm.Install('curl')
-    vm.RemoteCommand('curl %s -L -o %s/%s' % (
-        NETPERF_URL, INSTALL_DIR, NETPERF_TAR))
+def _LoadNetperf(vm):
+  vm.Install('curl')
+  vm.RemoteCommand('cd %s && git clone %s %s' % (
+      INSTALL_DIR, NETPERF_GIT, NETPERF_DIR))
```
Comment (s-deitz) [inline perfkitbenchmarker/linux_packages/netperf.py]: Is git clone getting a later version than we were getting with curl?

File: perfkitbenchmarker/linux_packages/netperf.py [hunk 7]
```diff
@@ -31,46 +30,45 @@
...
+                   './configure --enable-burst'
+                   '--enable-demo --enable-histogram '
+                   '&& make && sudo make install' %
+                   (NETPERF_DIR, FLAGS.netperf_histogram_buckets))
+
+  vm.RemoteCommand('cd %s && chmod +x runemomniaggdemo.sh'
+                   ' find_max_burst.sh'
+                   % (NETPERF_EXAMPLE_DIR))
 
-  Tries local data directory first, then NET_PERF_URL
```
Comment (s-deitz) [inline perfkitbenchmarker/linux_packages/netperf.py]: Why are we removing this functionality to get a local netperf tar file?

File: perfkitbenchmarker/linux_benchmarks/netperf_aggregate_benchmark.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+FLAGS = flags.FLAGS
+
+BENCHMARK_NAME = 'netperf_aggregate'
+BENCHMARK_CONFIG = """
+netperf_aggregate:
+  description: test packets per second performance
+  vm_groups:
+    vm_1:
+      vm_spec: *default_single_core
+    vm_2:
```
[the remark this answers is not in the document]
Comment (dPhanekham) [reply perfkitbenchmarker/linux_benchmarks/netperf_aggregate_benchmark.py]: changed to use client and server group with arbitrary number of server vms

File: perfkitbenchmarker/linux_benchmarks/netperf_aggregate_benchmark.py [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+# make  # sudo make install if you prefer
+  vm.Install('texinfo')
+  vm.Install('python_rrdtool')
+  vm.Install('netperf')
+
+  if vm.IS_REBOOTABLE:
+    vm.ApplySysctlPersistent({
+        'net.ipv4.tcp_keepalive_time': 60,
+        'net.ipv4.tcp_keepalive_intvl': 60,
+    })
```
[the remark this answers is not in the document]
Comment (dPhanekham) [reply perfkitbenchmarker/linux_benchmarks/netperf_aggregate_benchmark.py]: moved this to the netperf package. removed from this benchmark and netperf benchmark

File: perfkitbenchmarker/linux_benchmarks/netperf_aggregate_benchmark.py [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+  # Don't modify the metadata dict that was passed in
+
+  logging.info("Parsing netperf aggregate output")
+  metadata = metadata.copy()
+  aggregate_samples = []
+
+  stdout_ascii = stdout.encode("ascii")
+
+  for line in stdout_ascii.splitlines():
+    print(line)
```
[the remark this answers is not in the document]
Comment (dPhanekham) [reply perfkitbenchmarker/linux_benchmarks/netperf_aggregate_benchmark.py]: removed debugging statements

**Comment by dPhanekham** [comment]
I went back and changed quite a few things:

name changed to netperf_pps

changed it to use client and server vm groups, with an arbitrary number of servers

Instead of having a copy of the pps script in perfkitbenchmarker, I use the version already in netperf. A small patch file was already being used to make a few changes to netperf, so I added the changes I needed to make to the pps script to that patch file, as well as a small fix that needed to be made so that it would compile with the enable-demo flag.

**Comment by dPhanekham** [comment]
Yes. It can be run with the command ./pkb.py --benchmarks=netperf_pps
I can also provide a .yaml file and or the full output of the run if requested

**Comment by dPhanekham** [comment]
Here is a .yaml file for it. I can also add this or something similar as an example to the config directory

netperf_pps:
  description: test packets per second performance
  vm_groups:
    servers:
      vm_spec:
        GCP:
          zone: us-east4-a
          machine_type: n1-standard-2
      vm_count: 4
    client:
      vm_spec:
        GCP:
          zone: us-east4-a
          machine_type: n1-standard-2

**Pull request merged**